R-package/tests/testthat/test_lgb.Booster.R

```diff
@@ -799,37 +799,170 @@ test_that("all parameters are stored correctly with save_model_to_string()", {
         data = matrix(rnorm(500L), nrow = 100L)
         , label = rnorm(100L)
     )
-    nrounds <- 4L
     bst <- lgb.train(
         params = list(
-            objective = "regression"
-            , metric = "l2"
+            objective = "mape"
+            , metric = c("l2", "mae")
             , num_threads = .LGB_MAX_THREADS
+            , seed = 708L
+            , data_sample_strategy = "bagging"
+            , sub_row = 0.8234
         )
         , data = dtrain
-        , nrounds = nrounds
+        , nrounds = 3L
         , verbose = .LGB_VERBOSITY
     )
 
-    model_str <- bst$save_model_to_string()
-    params_in_file <- .params_from_model_string(model_str = model_str)
+    # entries whose values should reflect params passed to lgb.train()
+    non_default_param_entries <- c(
+        "[objective: mape]"
+        # 'l1' was passed in with alias 'mae'
+        , "[metric: l2,l1]"
+        , "[data_sample_strategy: bagging]"
+        , "[seed: 708]"
+        # this was passed in with alias 'sub_row'
+        , "[bagging_fraction: 0.8234]"
+        , "[num_iterations: 3]"
+    )
+
+    # entries with default values of params
+    default_param_entries <- c(
+        "[boosting: gbdt]"
+        , "[tree_learner: serial]"
+        , "[device_type: cpu]"
+        , "[data: ]"
+        , "[valid: ]"
+        , "[learning_rate: 0.1]"
+        , "[num_leaves: 31]"
+        , sprintf("[num_threads: %i]", .LGB_MAX_THREADS)
+        , "[deterministic: 0]"
+        , "[histogram_pool_size: -1]"
+        , "[max_depth: -1]"
+        , "[min_data_in_leaf: 20]"
+        , "[min_sum_hessian_in_leaf: 0.001]"
+        , "[pos_bagging_fraction: 1]"
+        , "[neg_bagging_fraction: 1]"
+        , "[bagging_freq: 0]"
+        , "[bagging_seed: 15415]"
+        , "[feature_fraction: 1]"
+        , "[feature_fraction_bynode: 1]"
+        , "[feature_fraction_seed: 32671]"
+        , "[extra_trees: 0]"
+        , "[extra_seed: 6642]"
```

Comment:
Possibly noob question but where are these seeds coming from? I see the defaults are 3, 2 and 6.
https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/include/LightGBM/config.h#L342
https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/include/LightGBM/config.h#L362
https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/include/LightGBM/config.h#L372

Reply:
Great question! I actually just learned this about LightGBM myself as part of working on this.

If you provide a value for `seed` in `params`, all other seeds are deterministically derived from it.

https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/src/io/config.cpp#L243-L252

Reply:
And here are the docs on that:

https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/include/LightGBM/config.h#L239-L244

So LightGBM will generate all these seeds based on `seed` if provided, then selectively override with any other values you passed in through `params`, via `Config::GetMembersFromString()`

https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/src/io/config.cpp#L266

And if neither of those things resulted in a value being set for a particular one, like `feature_fraction_seed`, then the hard-coded defaults you linked to are used.